src/Identity/IdentityServer/CustomTokenRequestValidator.cs

```diff
@@ -52,10 +53,23 @@ public async Task ValidateAsync(CustomTokenRequestValidationContext context)
         string[] allowedGrantTypes = { "authorization_code", "client_credentials" };
         if (!allowedGrantTypes.Contains(context.Result.ValidatedRequest.GrantType)
             || context.Result.ValidatedRequest.ClientId.StartsWith("organization")
-            || context.Result.ValidatedRequest.ClientId.StartsWith("installation"))
+            || context.Result.ValidatedRequest.ClientId.StartsWith("installation")
+            || context.Result.ValidatedRequest.Client.AllowedScopes.Contains(ApiScopes.ApiSecrets))
         {
+            // TODO: Is this the best approach?
+            if (context.Result.ValidatedRequest.Client.Properties.ContainsKey("encryptedPayload"))
```

Comment:
Maybe check if this is null too? the property will always exist for api keys, but we should really only attach it if it exists.

Reply:
Added a whitespace check!